tokio/src/runtime/scheduler/multi_thread/worker.rs

```diff
@@ -612,7 +616,29 @@ impl Core {
         if self.tick % worker.handle.shared.config.global_queue_interval == 0 {
             worker.inject().pop().or_else(|| self.next_local_task())
         } else {
-            self.next_local_task().or_else(|| worker.inject().pop())
+            let maybe_task = self.next_local_task();
+
+            if maybe_task.is_some() {
+                return maybe_task;
+            }
+
+            let cap = self.run_queue.capacity();
+
+            // The worker is currently idle, pull a batch of work from the
+            // injection queue. We don't want to pull *all* the work so other
+            // workers can also get some.
+            let n = cmp::min(
+                worker.inject().len() / worker.handle.shared.remotes.len() + 1,
+                cap,
+            );
+
+            let n = cmp::min(n, self.run_queue.max_capacity() / 2);
+
+            worker.inject().pop_n(n, |task| {
+                if self.run_queue.push_back(task).is_err() {
```

Comment:
Why not add a `push_back_batch`, which takes an iterator over the tasks? This would allow only doing the atomic operations once instead of for every task that is pushed. 
I added something like that in the `bwos` merge request.

Reply:
I avoided it for simplicity initially, but I can look into it now.

Reply:
maybe worth making this change in a follow-up branch?

Reply:
I applied it to this PR.